tamr_toolbox/utils/downstream.py

```diff
@@ -0,0 +1,245 @@
+    if include_dependencies_by_name and _is_unified_dataset(
+        dataset_name, all_unified_datasets=all_unified_datasets
+    ):
+        dependencies_suggested_by_name = set(
+            [d for d in all_tamr_datasets.keys() if d.startswith(dataset_name)]
+        )
+        LOGGER.debug(
+            f"Dataset {dataset_name} has {len(dependencies_suggested_by_name)} "
+            f"datasets with dependency suggested by name"
+        )
+        datasets_check = datasets_check.union(dependencies_suggested_by_name)
+        # Filter out datasets that have already been checked and the target dataset
+        datasets_check = datasets_check.difference(downstream_datasets.keys()).difference(
+            {dataset.name}
+        )
+
+    # add the target dataset to list of downstream datasets to avoid checking later
+    downstream_datasets[dataset_name] = dataset
+
+    # check further downstream for either dependencies suggested by name
+    # or dependencies that used to or is unified dataset
+    for d in datasets_check:
+        _find_downstream_datasets(
+            d,
+            include_dependencies_by_name=include_dependencies_by_name,
+            client=client,
+            all_tamr_datasets=all_tamr_datasets,
+            all_unified_datasets=all_unified_datasets,
+            downstream_datasets=downstream_datasets,
+        )
+
+    return downstream_datasets
+
+
+def _find_associated_projects(dataset: Dataset) -> List[str]:
+    """Return list of project_list that the dataset is part of.
+
+    Args:
+        dataset: The target dataset.
+
+    Returns:
+        List of Project IDs that the target `dataset` is part of.
+
+    """
+    project_list = set(
+        [step.project().resource_id for step in dataset.usage().usage.input_to_project_steps]
+    )
+    return list(project_list)
+
+
+def _count_dependencies(dataset: Dataset) -> int:
+    """Return number of dependencies of a dataset.
+
+    Args:
+        dataset: The target dataset.
+
+    Returns:
+        Number of dependencies of the target `dataset`.
+
+    """
+    dependencies = dataset.usage().dependencies
+    return len(dependencies)
+
+
+def _is_unified_dataset(
+    dataset_name: str,
+    *,
+    check_regex: bool = True,
+    regex=".*unified_dataset$",
+    all_unified_datasets: Optional[List[str]] = None,
+) -> bool:
+    """Check if a dataset is or used to be unified dataset.
+
+    Args:
+        dataset: The target dataset.
+        check_regex: A Boolean to determine whether to check for regex matching if a dataset is
+            not part of the pipeline (if its project has been deleted.) Default to True.
+        regex: A string of regex to check. Default to '.*unified_dataset$' which is Tamr's current
+            default unified dataset naming.
+
+    Returns:
+        A Boolean, if True, the target `dataset` is or used to be unified dataset.
+
+    """
+    # check project pipeline (only works if project has not been deleted)
```

Comment:
I think you need to populate the optional value here using:

if not all_unified_datasets:
    all_unified_datasets = _get_all_unified_datasets(dataset.client)

This also means that the input should be a Dataset object not the dataset_name

Reply:
I forgot to update docs for this function. I actually changed to make this parameter no longer optional, and used name instead of dataset objects so that I do not need to extract list of names every time I check a dataset. The list of unified dataset names is generated in `_find_downstream_datasets` method.